## Name: Ujjwal Chowdhury
## Assignment 1
## Reg. No.: B2130051
## MSc BDA

# Reading
Download the file and Read it using pickle as given below.

It is a python dict: 
    
    'text' is str object consists of all the sentences from all the volumes  
    'count' is the exact distict words present

In [9]:
import pandas as pd # For reading the dataset
from pathlib import Path #  For getting the directory path
import json # For saving the list of words and tokens
PATH = str(Path.cwd().parent)

# Reading the dataset

In [13]:
import pickle
with open(PATH + "/data/processed/dataset.pkl", 'rb') as f:
    dataset = pickle.load(f) 
dataset.keys(), dataset['count']

(dict_keys(['text', 'count']), 27301)

In [14]:
df = pd.DataFrame([dataset])

# Data Cleaning & Preprocessing

In [16]:
df['text'] = df['text'].str.replace('.',' ')
df['text'] = df['text'].str.replace(',',' ')
df['text'] = df['text'].str.replace('?',' ')
df['text'] = df['text'].str.replace('!',' ')
df['text'] = df['text'].str.replace(';',' ')
df['text'] = df['text'].str.replace(':',' ')
df['text'] = df['text'].str.replace('“',' ')
df['text'] = df['text'].str.replace('“',' ')
df['text'] = df['text'].str.replace('”',' ')
df['text'] = df['text'].str.replace('-','')
df['text'] = df['text'].str.replace('—','')
df['text'] = df['text'].str.replace("’","")

In [18]:
str = df['text'][0]

In [21]:
# Lets print a portion of the text
print(str[:500])

 Response To Welcome At the Worlds Parliament of Religions  Chicago 11th September  1893 Sisters and Brothers of America  It fills my heart with joy unspeakable to rise in response to the warm and cordial welcome which you have given us  I thank you in the name of the most ancient order of monks in the world  I thank you in the name of the mother of religions  and I thank you in the name of millions and millions of Hindu people of all classes and sects  My thanks  also  to some of the speakers o


# Creating the list of words by tokenization

In [22]:
identifiers = {}
idx = 0
for word in str.split():
    if word not in identifiers:
        identifiers[word] = idx
        idx += 1

list_values = [identifiers[word] for word in str.split()]

len(list_values)

1486756

In [23]:
import json
# Savings the list of tokens as a simple text file
with open(PATH + '/data/processed/list.txt', 'w') as f:
    f.write(json.dumps(list_values))

# Now read the file back into a Python list object
with open(PATH + '/data/processed/list.txt', 'r') as f:
    data = json.loads(f.read())

# FM Algorithm

In [9]:

print("Hash functions are defined as (a*x+b) %c, where x is an element of the set.")
inputCount = int(input("Enter the number of hash functions: "))
abcList = []

for i in range(inputCount):
    inputList = input("Enter the space-separated values of a, b and c: ").split(" ")
    abcList.append([int(i) for i in inputList])
finalCountsRecorded = []

for i in abcList:
    binElems = []
    for j in set(data):             #Evaluates the hash function, then converts it to binary. 
        binElems.append(str(bin((i[0]*j+i[1])%i[2])).split("b")[1])  #Appends binary output to a list
    greatestTrailing = 0             
    for k in binElems:             #Processes every element for that specific hash
        reversedCount = k[::-1]    
        count = 0
        for i in reversedCount:
            if(i=='1'):
                if(count>greatestTrailing):
                    greatestTrailing = count       #The greatest number of trailing zeros are established
                break
            else:
                count+=1
    finalCountsRecorded.append(2**greatestTrailing)      #The formula R = 2^r is applied, where R is number of elems,
                  #and r is max trailing zeros recorded
print("Counts recorded for each hash: ",finalCountsRecorded)
divider = inputCount//2
set1 = finalCountsRecorded[:divider]
set2 = finalCountsRecorded[divider:]
means = [sum(set1)/inputCount,sum(set2)/inputCount]      #Note: For demonstration purposes, we only have two sets here.
median = sum(means)/2             #    Thus, median is defined as (sum of means)/2 

print("Approximate number of elements from mean-median approximation: ",int(median))

Hash functions are defined as (a*x+b) %c, where x is an element of the set.
Enter the number of hash functions: 2
Enter the space-separated values of a, b and c: 12345 98765 1234567
Enter the space-separated values of a, b and c: 123 653 1234567
Counts recorded for each hash:  [131072, 16384]
Approximate number of elements from mean-median approximation:  36864


In [11]:
print('Number of unique words in the dataset is:', len(set(data)))

Number of unique words in the dataset is: 36077


# Results

Approximate unique words by FM Algorithm 36864

Number of actual unique words in the dataset is: 36077